File: launcher/xpi-dl.go
package main

import (
	"archive/zip"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
)

// Downloads and extracts the default addons
func addDefaultAddons(excludeList []string, addonsList *[]string) {
	// Build a map from DefaultAddons, excluding keys found in excludeAddonsList
	addonsMap := make(map[string]string)
	for name, url := range DefaultAddons {
		if len(excludeList) == 0 || !contains(excludeList, name) {
			addonsMap[name] = url
		}
	}

	// Download if not already downloaded
	maybeDownloadAddons(addonsMap, addonsList)
}

// Downloads and extracts the addon
func downloadAndExtract(url, extractPath string) error {
	// Create a temporary file to store the downloaded zip
	tempFile, err := os.CreateTemp("", "camoufox-addon-*.zip")
	if err != nil {
		return fmt.Errorf("failed to create temp file: %w", err)
	}
	defer os.Remove(tempFile.Name()) // Clean up the temp file when done

	// Download the zip file
	resp, err := http.Get(url)
	if err != nil {
		return fmt.Errorf("failed to download addon: %w", err)
	}
	defer resp.Body.Close()

	// Write the body to the temp file
	_, err = io.Copy(tempFile, resp.Body)
	if err != nil {
		return fmt.Errorf("failed to write addon to temp file: %w", err)
	}

	// Close the file before unzipping
	tempFile.Close()

	// Extract the zip file
	err = unzip(tempFile.Name(), extractPath)
	if err != nil {
		return fmt.Errorf("failed to extract addon: %w", err)
	}

	return nil
}

// Extracts the zip file
func unzip(src, dest string) error {
	r, err := zip.OpenReader(src)
	if err != nil {
		return err
	}
	defer r.Close()

	for _, f := range r.File {
		fpath := filepath.Join(dest, f.Name)

		if f.FileInfo().IsDir() {
			os.MkdirAll(fpath, os.ModePerm)
			continue
		}

		if err = os.MkdirAll(filepath.Dir(fpath), os.ModePerm); err != nil {
			return err
		}

		outFile, err := os.OpenFile(fpath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, f.Mode())
		if err != nil {
			return err
		}

		rc, err := f.Open()
		if err != nil {
			outFile.Close()
			return err
		}

		_, err = io.Copy(outFile, rc)
		outFile.Close()
		rc.Close()

		if err != nil {
			return err
		}
	}
	return nil
}

// Checks if a slice contains a string
func contains(slice []string, item string) bool {
	for _, s := range slice {
		if s == item {
			return true
		}
	}
	return false
}

// Returns the absolute path to the target addon location
func getAddonPath(addonName string) string {
	return getPath(filepath.Join("addons", addonName))
}

// Downloads and extracts the addons
func maybeDownloadAddons(addons map[string]string, addonsList *[]string) {
	for addonName, url := range addons {
		// Get the addon path
		addonPath := getAddonPath(addonName)

		// Check if the addon is already extracted
		if _, err := os.Stat(addonPath); !os.IsNotExist(err) {
			// Add the existing addon path to addonsList
			*addonsList = append(*addonsList, addonPath)
			continue
		}

		// Addon doesn't exist, create directory and download
		err := os.MkdirAll(addonPath, 0755)
		if err != nil {
			fmt.Printf("Failed to create directory for %s: %v\n", addonName, err)
			continue
		}

		err = downloadAndExtract(url, addonPath)
		if err != nil {
			fmt.Printf("Failed to download and extract %s: %v\n", addonName, err)
		} else {
			fmt.Printf("Successfully downloaded and extracted %s\n", addonName)
			// Add the new addon directory path to addonsList
			*addonsList = append(*addonsList, addonPath)
		}
	}
}

File: pythonlib/camoufox/xpi_dl.py
import os
from multiprocessing import Lock
from typing import List, Optional

from .addons import DefaultAddons
from .pkgman import get_path, unzip, webdl


def add_default_addons(
    addons_list: List[str], exclude_list: Optional[List[DefaultAddons]] = None
) -> None:
    """
    Adds default addons, minus any specified in exclude_list, to addons_list
    """
    # Build a dictionary from DefaultAddons, excluding keys found in exclude_list
    if exclude_list is None:
        exclude_list = []

    addons = [addon for addon in DefaultAddons if addon not in exclude_list]

    with Lock():
        maybe_download_addons(addons, addons_list)


def download_and_extract(url: str, extract_path: str, name: str) -> None:
    """
    Downloads and extracts an addon from a given URL to a specified path
    """
    # Create a temporary file to store the downloaded zip
    buffer = webdl(url, desc=f"Downloading addon ({name})", bar=False)
    unzip(buffer, extract_path, f"Extracting addon ({name})", bar=False)


def get_addon_path(addon_name: str) -> str:
    """
    Returns a path to the addon
    """
    return get_path(os.path.join("addons", addon_name))


def maybe_download_addons(
    addons: List[DefaultAddons], addons_list: Optional[List[str]] = None
) -> None:
    """
    Downloads and extracts addons from a given dictionary to a specified list
    Skips downloading if the addon is already downloaded
    """
    for addon in addons:
        # Get the addon path
        addon_path = get_addon_path(addon.name)

        # Check if the addon is already extracted
        if os.path.exists(addon_path):
            # Add the existing addon path to addons_list
            if addons_list is not None:
                addons_list.append(addon_path)
            continue

        # Addon doesn't exist, create directory and download
        try:
            os.makedirs(addon_path, exist_ok=True)
            download_and_extract(addon.value, addon_path, addon.name)
            # Add the new addon directory path to addons_list
            if addons_list is not None:
                addons_list.append(addon_path)
        except Exception as e:
            print(f"Failed to download and extract {addon.name}: {e}")
